src/KubernetesClient/Controllers/ResourceEventDeltaBlock.cs

```diff
@@ -0,0 +1,297 @@
+using System;
+using System.Collections.Generic;
+using System.Linq;
+using System.Reactive.Disposables;
+using System.Threading.Tasks;
+using System.Threading.Tasks.Dataflow;
+using k8s.Informers.Notifications;
+using Microsoft.Extensions.Logging;
+using Microsoft.Extensions.Logging.Abstractions;
+using Org.BouncyCastle.Asn1.X509;
+
+namespace k8s.Controllers
+{
+
+	/// <summary>
+	/// A TPL Dataflow block exposes queuing semantics that attaches to infromer observable streams.
+	/// It groups delta changes per key, sending each group to downstream blocks as individual messages
+	/// This allows "Per resource ID" processing semantics
+	/// Any sync events are only propagated if no other changes are queued up
+	/// </summary>
+	/// <typeparam name="TKey">The type of key used to identity resource</typeparam>
+	/// <typeparam name="TResource">The resource type</typeparam>
+	public class ResourceEventDeltaBlock<TKey, TResource> : ISourceBlock<List<ResourceEvent<TResource>>>, ITargetBlock<ResourceEvent<TResource>>
+	{
+		private readonly Func<TResource, TKey> _keyFunc;
+		private readonly bool _skipTransient;
+        private readonly ILogger _log;
+        private readonly Queue<List<ResourceEvent<TResource>>> _queue = new Queue<List<ResourceEvent<TResource>>>();
+		private readonly List<TargetLink> _targets = new List<TargetLink>();
+		private bool _isCompleting;
+		private long _msgId;
+		private readonly object _lock = new object();
```

Comment:
Please document what this lock protects, it will make it easier to review/understand the code.